pkg/controller/operators/cleanupcsv_controller.go

```diff
@@ -0,0 +1,151 @@
+)
+
+// CleanupCsvReconciler cleans up cluster-scoped CSV-owned resources.
+type CleanupCsvReconciler struct {
+	client.Client
+
+	log     logr.Logger
+	factory decorators.OperatorFactory
+}
+
+// SetupWithManager adds the operator reconciler to the given controller manager.
+func (r *CleanupCsvReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&operatorsv1alpha1.ClusterServiceVersion{}).
+		Complete(reconcile.Func(r.ReconcileClusterServiceVersion))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewCleanupCsvReconciler constructs and returns an CleanupCsvReconciler.
+// As a side effect, the given scheme has operator discovery types added to it.
+func NewCleanupCsvReconciler(cli client.Client, log logr.Logger, scheme *runtime.Scheme) (*CleanupCsvReconciler, error) {
+	// Add watched types to scheme.
+	if err := AddToScheme(scheme); err != nil {
+		return nil, err
+	}
+
+	factory, err := decorators.NewSchemedOperatorFactory(scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CleanupCsvReconciler{
+		Client: cli,
+
+		log:     log,
+		factory: factory,
+	}, nil
+}
+
+// ReconcileClusterServiceVersion projects the component labels of a given CSV onto all resources owned by it.
+func (r *CleanupCsvReconciler) ReconcileClusterServiceVersion(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
+	// Set up a convenient log object so we don't have to type request over and over again
+	log := r.log.WithValues("request", req)
+	log.V(1).Info("reconciling csv")
+	metrics.EmitCleanupCSVReconcile(req.Namespace, req.Name)
+
+	// Fetch the CSV from the cache
+	in := &operatorsv1alpha1.ClusterServiceVersion{}
+	if err := r.Get(ctx, req.NamespacedName, in); err != nil {
+		if apierrors.IsNotFound(err) {
+			err = nil
+		} else {
+			log.Error(err, "Error finding ClusterServiceVersion")
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Ignore copied csvs
+	if in.IsCopied() {
+		return reconcile.Result{}, nil
+	}
+
+	myFinalizerName := "olm.csv.cleanup.controller/finalizer"
+	if in.ObjectMeta.DeletionTimestamp.IsZero() {
+		// not being deleted, add finalizer if not present
+		if !controllerutil.ContainsFinalizer(in, myFinalizerName) {
+			controllerutil.AddFinalizer(in, myFinalizerName)
+			if err := r.Update(ctx, in); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		// csv is being deleted
+		if controllerutil.ContainsFinalizer(in, myFinalizerName) {
+			// finalizer still present, do any clean up
+			ownerSelector := ownerutil.CSVOwnerSelector(in)
+			deleteOptions := client.DeleteAllOfOptions{
+				ListOptions: client.ListOptions{
+					LabelSelector: ownerSelector,
+				},
+			}
+			// Look for resources owned by this CSV, and delete them.
+			log.WithValues("selector", ownerSelector).Info("Cleaning up resources after CSV deletion")
+
+			err := r.Client.DeleteAllOf(ctx, &rbacv1.ClusterRoleBinding{}, &deleteOptions)
```

Comment:
Probably arguments either way, but it may make more sense to attempt to send all of these calls even if one fails, instead of bailing on the first error. WDYT?

Reply:
Certainly an option, and I think I may have done that in another iteration.